src/video.js

```diff
@@ -35,15 +35,16 @@ export const hasNonVideoBidder = adUnit =>
 export function isValidVideoBid(bid, {index = auctionManager.index} = {}) {
   const videoMediaType = deepAccess(index.getMediaTypes(bid), 'video');
   const context = videoMediaType && deepAccess(videoMediaType, 'context');
+  const useCacheKey = videoMediaType && deepAccess(videoMediaType, 'useCacheKey');
   const adUnit = index.getAdUnit(bid);
 
   // if context not defined assume default 'instream' for video bids
   // instream bids require a vast url or vast xml content
-  return checkVideoBidSetup(bid, adUnit, videoMediaType, context);
+  return checkVideoBidSetup(bid, adUnit, videoMediaType, context, useCacheKey);
 }
 
-export const checkVideoBidSetup = hook('sync', function(bid, adUnit, videoMediaType, context) {
-  if (videoMediaType && context !== OUTSTREAM) {
+export const checkVideoBidSetup = hook('sync', function(bid, adUnit, videoMediaType, context, useCacheKey) {
+  if (videoMediaType && (useCacheKey || context !== OUTSTREAM)) {
```

Comment:
I am missing some context - why were we disabling video cache for outstream? I'm trying to understand if it'd make more sense to have this as a global toggle that says "enable caching for all video", rather than a flag for each adUnit. maybe `setConfig({cache: {outstream: true}})`?

If we do a setting at the adUnit level:

 - I think `cache` or `useCache` would be a better name than `useCacheKey` 
 - do we need to keep the distinction between outstream and the rest? that is, could this say `if (videoMediaType && useCache)`, with `useCache` defaulting to false for outstream, and true otherwise?

Reply:
I don't really know why this is coded for instream only, but I use this in my fork for month right now, so I can pass a dynamic url to a vast wrapper for counting. Maybe earlier this was thought to be needed as you always needed a "renderer" to setup outstream? I don't know, but to not break other things I thought it would be good to keep the actual state and only add a new option.

I'm also fine to set this as a global option, maybe even have both for most flexibility, while adUnit setting would overwrite the global. 

And if we go on here I will be pleased to change useCacheKey to whatever you want.

Reply:
We tried to investigate and it's unfortunately still unclear why we originally decided to not cache outstream. @jsnellbaker do you happen to remember some context? it seems that outstream caching was disabled in https://github.com/MikeSperone/Prebid.js/commit/f272031f1fc5f44c08ac7e0ebee5e47bf973dbe9#diff-e9f987f170d6e2065c627eb6a145180770dc0921cb2c82f0c0a549ae87934a4dR363

Reply:
I've found this:
https://github.com/prebid/Prebid.js/pull/3024#discussion_r216731281
and this:
https://github.com/prebid/Prebid.js/issues/6078

but would adding using cache for outstream as an optin to the pageOptions affect this?
I do need the cache to convert a xml response to an url to pass it as a query parameter (to our adserver).

Reply:
From what I always understood, video caching was utilized for setups where a video bid was being fed into a player integrated onto the page, since you didn't know when exactly the video bid was going to be rendere (ie pre-roll, mid-roll, post-roll, etc).  I may be mistaken (or out of date), but I think there was some ttl issues where the video bid could expire if it was directly accessed from the prebid.js window instance from the player.  So this led to having the vast creative info being cached (with a longer shelf-life) and pulled later at the player's convenience.

Outstream though is generally assumed to A) come with its own renderer (either supplied with the bid itself from the bidder's adserver or from the publisher setup in the adUnit) and B) the ad experience is simpler than instream (where you may just see the ad in the outstream player and that's it).  Additionally, outstream placements generally have their own dedicated section on the page, much like any other banner placement - so the ad could be rendered immediately as soon as the publisher's adserver (eg GAM) decided to choose that outstream prebid bid.  Therefore, there wasn't a need to cache things - since the bid information would typically be used asap and it had the means to do its rendering on its own.

This is my understanding - which I was not an original author on the initial outstream/instream workflow for prebid and I only learned from some others as they knew it.  

All that being said, as for why we can't use caching with outstream - the above basically distills to a 'it wasn't done like that before' (at least that's what I see looking over it again).  If there are needs now to support it - I'm not opposed to the changes and there may be a way to make things 'work', though there would need to be extensive thought and testing to handle the old-way in conjunction with this new way - since I don't think most outstream setups would support caching or see a need to move in that direction (unless there were strong advantages).

Reply:
I prefer simplicity, and that would be enough of a reason for me to use caching without introducing a distinction between instream and outstream.  Maybe it's not often necessary but that doesn't seem enough of a reason to forbid it in code. It sounds like you are in that situation @renebaudisch? You have a player set up to work with cache URLs and it'd be nice to not have to set up a whole other process.

Reply:
I have a video setup that works with urls and expects a xml vast response. It is used on the floating videoplayer e.g. on this url: https://www.computerbild.de/artikel/cb-Ratgeber-Software-kostenloser-Virenschutz-1153794.html

First we're using prebid. The winner get's passed to xandr for the rtb auction done by a ptv request.
So, if now the prebid bid returns as a winner, the player will need a xml wrapper containing an url coming from the adserver as response, this is where the cacheKey url comes into play.

As we need to host this response in Xandr as a video creative we cannot return the prebid xml by ourselves, even if we would pass it through. And also, as the vast plugin already is working, we also cannot interfere here and replace it.

Sure, we also could change the context to instream, but as this is whether in- nor outstream and we need a way to separate small view from fullview to the bidders, I'm back at the point where I need the cacheKey url...
Because, the small view is sending outstream context while the fullview on click sends instream to show the bidder what they are bidding for.

And since my change is backwards compatible I though why not submit it.